Add ability to compute auxiliary quantities from the mortar mesh

lindsayad opened the pull request:
### Example

From the solution continuity test which has a gap between secondary and primary bodies of .01

<img width="878" alt="Screen Shot 2022-01-11 at 3 10 11 PM" src="https://user-images.githubusercontent.com/10248304/149035929-b41ddd23-f197-4814-bc10-4bf8602e8276.png">

Closes #20001

lindsayad committed:
mortar nodal aux kernel

lindsayad committed:
factor out reinit preparation

lindsayad committed:
Move material property APIs down from SubProblem to FEProblemBase

All the material storage in in FEProblemBase. It is not very safe to
have the material property APIs down in SubProblem because a user of the
APIs could call on the displaced problem and then when it is time to
reinit the materials, it's possible we check and see that there are no
active material properties on the FEProblemBase and then we don't reinit

lindsayad committed:
Working demonstration with WeightedGapAux

Includes test

Closes #20001

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/929135/) on 71fcbca wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/20017/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/20017/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format a6076d579c4ea10b0c3bdb7d7561334faf24c241`

recuero commented on framework/src/auxkernels/MortarNodalAuxKernel.C:
```diff
@@ -0,0 +1,75 @@
+    total_volume += _msm_volume;
+  };
+
+  Moose::Mortar::loopOverMortarSegments(
+      its, _assembly, _subproblem, _fe_problem, amg(), _displaced, consumers, act_functor);
```
This method is computing the nodal value (using the `act_functor` among others): Do you think its name should be changed to reflect this? Say, something like `computeValueLoopingOverMortarSegments`. Maybe the answer is just that name is too long.

lindsayad replied:
This particular call to `loopOverMortarSegments` will `computeValue` for an aux kernel. But in `ComputeMortarFunctor` we call `loopOverMortarSegments` with an action functor that calls `computeResidual` or `computeJacobian`. The flexibility in `loopOverMortarSegments` is that we do not know what the user of the function will be doing in their action functor

recuero replied:
Makes sense.

recuero commented on framework/src/loops/ComputeMortarFunctor.C:
```diff
@@ -100,200 +66,48 @@ ComputeMortarFunctor::operator()()
-  }
+  };
+
+  Moose::Mortar::loopOverMortarSegments(iterators,
+                                        _assembly,
+                                        _subproblem,
+                                        _fe_problem,
+                                        _amg,
+                                        _displaced,
+                                        _mortar_constraints,
+                                        act_functor);
```
Nice!

lindsayad replied:
😄

recuero commented on framework/include/utils/MortarUtils.h:
```diff
@@ -0,0 +1,251 @@
+                      const QBase & qrule_msm,
+                      std::vector<Point> & q_pts);
+
+template <typename Iterators, typename Consumers, typename ActionFunctor>
+void
+loopOverMortarSegments(const Iterators & secondary_elems_to_mortar_segments,
+                       Assembly & assembly,
+                       SubProblem & subproblem,
+                       FEProblemBase & fe_problem,
+                       const AutomaticMortarGeneration & amg,
+                       const bool displaced,
+                       const Consumers & consumers,
+                       const ActionFunctor act)
```
What's the optimization resulting from leaving this implementation in the header file?

lindsayad replied:
Probably no optimization/inlining, but template function definitions generally need to be in the header unless you do explicit template instantiations in the .C file

recuero commented on test/tests/mortar/aux-gap/gap.i:
```diff
@@ -0,0 +1,123 @@
+  [../]
+[]
+
+[AuxVariables]
+  [gap]
+    block = '10'
+  []
+[]
+
+[AuxKernels]
+  [gap]
+    type = WeightedGapAux
+    variable = gap
+    primary_boundary = 2
+    secondary_boundary = 1
+    primary_subdomain = 20
+    secondary_subdomain = 10
+    block = '10'
+  []
+[]
```
What happens code-wise if the aux variable interpolation order doesn't coincide with that of the primal variables?

lindsayad replied:
See the new test I added. The answer is that it doesn't matter. Since we are doing evaluations on mortar segments where finite element spaces are well defined for every quantity used in the aux computation, we should always be good to go. So match/mismatch with other variable finite element families/orders doesn't matter. You can always use a nodal aux variable that is of order one up to the order of the mesh (so really either order 1 or 2)

recuero reviewed (commented):
Awesome work!

recuero commented on framework/include/problems/SubProblem.h:
```diff
@@ -288,39 +288,6 @@ class SubProblem : public Problem
    */
   virtual void clearActiveElementalMooseVariables(THREAD_ID tid);
 
-  /**
-   * Record and set the material properties required by the current computing thread.
-   * @param mat_prop_ids The set of material properties required by the current computing thread.
-   *
-   * @param tid The thread id
-   */
-  virtual void setActiveMaterialProperties(const std::set<unsigned int> & mat_prop_ids,
-                                           THREAD_ID tid);
-
```
Never mind. Explained in the commit message.

lindsayad replied:
I'm copying this from one of my commit messages:
> All the material storage in in FEProblemBase. It is not very safe to
have the material property APIs down in SubProblem because a user of the
APIs could call on the displaced problem and then when it is time to
reinit the materials, it's possible we check and see that there are no
active material properties on the FEProblemBase and then we don't reinit the 
materials that are *only* ever held by `FEProblemBase`

Is that a clear enough explanation? A "user of the APIs" happened to be me when I was trying to convert all calls on `fe_problem` to `subproblem`

recuero replied:
Yeah, I think it's clear and that the change appears to make problem API usage more robust. I didn't see the description that you pasted initially.

lindsayad replied:
didn't reload the page, so I didn't see your second comment before I made mine! 😄

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/929291/) on 9a03503 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/20017/site)

This comment will be updated on new commits.

lindsayad commented on framework/src/auxkernels/MortarNodalAuxKernel.C:
```diff
@@ -0,0 +1,75 @@
+
+  // We have to reinit the node for this variable in order to get the dof index set for the node
+  _var.reinitNode();
+  _var.setNodalValue(value / total_volume);
```
For the weighted gap for example we multiply by JxW in the numerator, so eventually we need to divide by the summation of JxW in the denominator

lindsayad committed:
Documentation

lindsayad commented on framework/include/constraints/AutomaticMortarGeneration.h:
```diff
@@ -300,103 +300,126 @@ class AutomaticMortarGeneration : public ConsoleStreamInterface
+   * @return the lower dimensional secondary elements and their associated mortar segment elements
+   * that would have nonzero values for a Lagrange shape function associated with the provided node
+   */
+  std::vector<MortarFilterIter> secondariesToMortarSegments(const Node & node) const;
```
I've revised. See whether it makes more sense now

recuero replied:
Yeah, and thanks for adding those doxygen comments for the utility method.

lindsayad committed:
Add test in which aux is of less order than mesh

lindsayad committed:
Slightly loosen tolerance on contact test

I'm really not sure what change would have caused this small diff but it
doesn't look like a meaningful diff

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/930670/) on 1361991 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/20017/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/20017/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format a2ae9d093a927ca4dbf280884cecd5805f450e37`

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/930773/) on 0bf87d0 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>a2ae9d</th>
      <th colspan=4>#20017 0bf87d</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>82.19%</td>
      <td>82.20%</td>
      <td>+0.01%</td>
      <td>91.50%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>72963</td>
      <td>73036</td>
      <td>+73</td>
      <td>323</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15808</td>
      <td>15817</td>
      <td>+9</td>
      <td>30</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/20017/coverage_diff/framework.html)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/20017/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/20017/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/20017/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/20017/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/20017/coverage/contact)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/20017/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/20017/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/20017/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/20017/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/20017/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/20017/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/20017/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/20017/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/20017/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/20017/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/20017/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/20017/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/20017/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/20017/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/20017/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/20017/coverage/richards)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/20017/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/20017/coverage/tensor_mechanics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/20017/coverage/xfem)

</details>



This comment will be updated on new commits.

lindsayad commented:
Thank you to the nonsparse AD test with `-p4 --distributed-mesh` that was the only recipe which caught a major issue with this design ... this current design will not work with the current level of ghosting for mortar ... doing node-based loops requires that we would expand our ghosting pattern to include all mortar segments that contribute to process-local nodes connected to secondary elements. This is different from `MortarConstraints` which only require mortar segments for process-local secondary elements.

lindsayad committed:
Don't prepare_for_use in mortar segment mesh

We could potentially need remote mortar elements for our new
`MortarNodalAuxKernel`. This need would arise any time a node is on a
processor boundary and consequently the aux kernel would call for
evaluations of non-local mortar segments associated with that node. And
if we are not deleting remote elements, partitioning, or finding
neighbors, then I don't really know why we should call `prepare_for_use`

lindsayad committed:
Add option to ghost point neighbors for mortar

lindsayad committed:
Add all second order test

lindsayad committed:
Gruesome wrestling with ghosting but I think we've emerged victorious

lindsayad committed:
Remove non-grid-sequencing version of test

This test is challenging enough that the non-grid-sequencing version of
it has a solve that is not robust and can occassionally fail just by
changing the executable you run with. As removing this test will not
decrease code coverage, I think removing it is the right thing to do

recuero commented:
>Gruesome wrestling with ghosting but I think we've emerged victorious

😆

recuero commented on framework/src/relationshipmanagers/AugmentSparsityOnInterface.C:
```diff
@@ -227,21 +227,70 @@ AugmentSparsityOnInterface::operator()(const MeshBase::const_element_iterator &
     {
       ghost_mortar_interface_couplings(elem);
 
-      if (_ghost_point_neighbors && (elem->subdomain_id() == secondary_subdomain_id))
```
Can't understand why the second condition is removed.

lindsayad replied:
It's related to this comment:
```c++
// I hypothesize that node processor ids are tied to higher dimensional element processor
// ids over lower dimensional element processor ids based on debugging experience.
// Consequently we need to be checking for higher dimensional elements and whether they are
// along our secondary boundary. From there we will query the AMG object's
// mortar-interface-coupling container to get the secondary lower-dimensional element, and
// then we will ghost it's point neighbors and their mortar interface couples
```
Essentially we are going to initiate our search for point neighbors starting with higher dimensional elements along the secondary boundary, *not* from lower dimensional elements along the secondary subdomain

lindsayad replied:
Node processor ids are what are important for our mortar nodal aux kernel evaluations

recuero commented on modules/contact/test/tests/3d-mortar-contact/tests:
```diff
@@ -83,7 +83,7 @@
                   'auxiliary kernel through a user-friendly action. Results are diffed against '
                   'non-action output'
-    rel_err = 1.0e-5
+    rel_err = 3.0e-5
```
Do we still need this after your expanded ghosting  work?

lindsayad replied:
Most likely. We do not ghost point neighbors unless one of our mortar nodal aux kernels is present, so in theory the ghosting in this particular simulation should be totally unchanged.

lindsayad committed:
Remove need to build secondary_to_msms structure in ComputeMortarFunctor

We now do not add small volume mortar segment elements to the 2D mortar
mesh, which is now consistent with what we do in 3D. With this change we
don't have to check for small volume mortar segments in
ComputeMortarFunctor. We can just build our container of iterators based
on whether the secondary element is local and whether the mortar
segment container associated with it is non-empty

lindsayad committed:
Make AugmentSparsityOnInterface thread safe

recuero commented on framework/include/relationshipmanagers/AugmentSparsityOnInterface.h:
```diff
@@ -75,13 +75,6 @@ class AugmentSparsityOnInterface : public RelationshipManager
 protected:
   virtual void internalInitWithMesh(const MeshBase &) override;
 
-  /**
-   * The Mesh we're calculating on
-   */
-  const AutomaticMortarGeneration * _amg;
-
-  bool _has_attached_amg;
-
```
how do this commit's changes address thread safety?

lindsayad replied:
We were reading from and writing to the `_amg` pointer and the `_has_attached_amg` boolean, which are shared pieces of data across threads, without any protection. So there would be data races for those. By making the `amg` pointer a stack variable, each thread now has it's own copy of the pointer and there are no races for reading/writing

recuero reviewed (approved):
Nice cleanup! I'll test this on non-flat surfaces too.

lindsayad commented:
Anybody else want to look at this?
